MaxText/inference/page_manager.py

```diff
@@ -1,232 +1,582 @@
+  current_num_pages_used = released_state.num_pages_used
+
+  def handle_zero_length(_unused_operand) -> PageState:
+    """Handles the case where true_length is 0."""
+    return released_state
+
+  def handle_normal_case(_unused_operand) -> PageState:
+    """Handles the case where true_length > 0."""
+    num_pages_needed = (true_length + tokens_per_page - 1) // tokens_per_page
+    last_page_position = (true_length - 1) % tokens_per_page
+
+    num_free_pages = jnp.sum(current_page_status == 0)
+    group_has_capacity = num_pages_needed <= max_pages_per_group
+    has_enough_resources = (int(num_free_pages) >= num_pages_needed) and group_has_capacity
+
+    def allocate_and_update_state(initial_state_tuple: Tuple[PagesInt1d, GroupsPagesInt2d, GroupsInt1d]) -> PageState:
+      """Allocates pages if resources are sufficient."""
+
+      def allocate_one_page(
+          page_idx_in_group: int, loop_state_tuple: Tuple[PagesInt1d, GroupsPagesInt2d, GroupsInt1d]
+      ) -> Tuple[PagesInt1d, GroupsPagesInt2d, GroupsInt1d]:
+        """Allocates a single page within the fori_loop."""
+        current_loop_status, current_loop_map, current_loop_num_used = loop_state_tuple
+
+        next_free_page_global = _find_next_free_page_index(current_loop_status)
+        page_allocated = next_free_page_global >= 0
+
+        new_loop_status = jax.lax.cond(
+            page_allocated, lambda s: s.at[next_free_page_global].set(1), lambda s: s, current_loop_status
+        )
+        new_loop_map = jax.lax.cond(
+            page_allocated,
+            lambda m: m.at[page_group_id, page_idx_in_group].set(next_free_page_global),
+            lambda m: m,
+            current_loop_map,
+        )
+        new_loop_num_used = jax.lax.cond(
+            page_allocated, lambda n: n.at[page_group_id].add(1), lambda n: n, current_loop_num_used
+        )
+        return new_loop_status, new_loop_map, new_loop_num_used
+
+      final_page_status, final_page_map, final_num_pages_used = jax.lax.fori_loop(
+          0, num_pages_needed, allocate_one_page, initial_state_tuple
+      )
+
+      active_page_global_index = final_page_map[page_group_id, num_pages_needed - 1]
+
+      return released_state.replace(
+          page_status=final_page_status,
+          page_map=final_page_map,
+          num_pages_used=final_num_pages_used,
+          sequence_lengths=released_state.sequence_lengths.at[page_group_id].set(true_length),
+          active_page=released_state.active_page.at[page_group_id].set(active_page_global_index),
+          has_active_page=released_state.has_active_page.at[page_group_id].set(True),
+          active_page_position=released_state.active_page_position.at[page_group_id].set(last_page_position),
+      )
+
+    if has_enough_resources:
+      return allocate_and_update_state((current_page_status, current_page_map, current_num_pages_used))
+    else:
+      return released_state
+
+  if true_length == 0:
```

Comment:
Use python native condition when the value is known at compile time. 

In this case, true length is known only at computation time, so we may want to use jax.lax.cond

Reply:
Good catch, thought I had caught all of these. Will change.